Add spring arm third-person camera tutorial (#10156)
godotengine/godot-docs: tutorials/3d/spring_arm.rst

:article_outdated: False

.. _doc_spring_arm:

Third-person camera with spring arm
===================================

.. video:: video/spring_arm_camera.webm
   :alt: Camera attached to a spring arm colliding with walls
   :autoplay:
   :loop:

>    :muted:

Reviewer: Why is this same video at the top *and* bottom of the page?

I would probably just put it at the bottom, since it shows off the final version.

Author: I had also put it at that top to show what the tutorial will end with, but just having it at the bottom is fine too 👍

3D games will often have a third-person camera that follows and
rotates around something such as a player character or a vehicle.

In Godot, this can be done by setting a :ref:`Camera3D <class_Camera3D>` as a child of a node.
However, if you try this without any extra steps you'll notice that the camera clips through geometry and hides the scene.

This is where the :ref:`SpringArm3D <class_SpringArm3D>` node comes in.

What is a spring arm?
---------------------

A spring arm has two main components that affect its behavior.

The "length" of the spring arm is how far from its global position to check for collisions:

.. image:: img/spring_arm_position_length.svg

The "shape" of the spring arm is what it uses to check for collisions. The spring arm will "sweep" this shape from its origin out towards its length.

.. image:: img/spring_arm_shape.svg

The spring arm tries to keep all of its children at the end of its length. When the shape collides with something, the children are instead placed at or near that collision point:

.. image:: img/spring_arm_children.svg

> Spring arm with a camera

Reviewer: This whole section looks good.

------------------------

When a camera is placed as a child of a spring arm, a pyramid representing the camera will be used as the shape.

This pyramid represents the **near plane** of the camera:

.. image::: img/spring_arm_camera_shape.svg

.. note:: If the spring arm is given a specific shape, then that shape will **always** be used.

    The camera's shape is only used if the camera is a **direct child** of the spring arm.

    If no shape is provided and the camera is not a direct child, the spring arm will fall back to using a ray cast which is inaccurate for camera collisions and not recommended.

Every physics process frame, the spring arm will perform a motion cast to check if anything is collided with:

.. image:: img/spring_arm_camera_motion_cast.svg

When the shape hits something, the camera will be placed at the collision point:

.. image:: img/spring_arm_camera_collision.svg

Setting up the spring arm and camera
------------------------------------

Let's add a spring arm camera setup to the platformer demo.

> .. note:: You can download the Platformer3D demo on `GitHub <https://github.com/godotengine/godot-demo-projects/tree/master/3d/platformer>`_ or using the `Asset Library <https://godotengine.org/asset-library/asset/2748>`_.

Reviewer: ```suggestion
.. note:: You can download the Platformer 3D Demo on `GitHub <https://github.com/godotengine/godot-demo-projects/tree/master/3d/platformer>`_ or using the `Asset Library <https://godotengine.org/asset-library/asset/2748>`_.
```
Use the actual name of the demo. I tried searching for "Platformer3D" and nothing came up

Author: Good catch! It also needs changed in these two spots (where I copied it from 🫣 )

https://github.com/godotengine/godot-docs/blob/84842ebe00b9b60f540349f62e92c7ad10ac74bf/tutorials/physics/ragdoll_system.rst?plain=1#L11-L13

https://github.com/godotengine/godot-docs/blob/84842ebe00b9b60f540349f62e92c7ad10ac74bf/tutorials/physics/soft_body.rst?plain=1#L35-L37

Reviewer: Might have been called something else on the asset library page when these were originally written

In general, for a third-person camera setup, you will have three nodes as children of the node that you're following:

- `Node3D` (the "pivot point" for the camera)

    - `SpringArm3D`

        - `Camera3D`

> Set these up as children of our player and give them unique names so we can find them in our script:

Reviewer: ```suggestion
Open the ``player.tscn`` scene. Set these up as children of our player and give them unique names so we can find them in our script:
```
If we're meant to follow along, there should be some specific instruction on which file to open.

.. image:: img/spring_arm_editor_setup.png

> Let's move the pivot point up by `1.5` on the Y-axis so that it's not on the ground:

Reviewer: ```suggestion
Let's move the pivot point up by ``1.5`` on the Y-axis so that it's not on the ground:
```
Literal values use `code` style, which is done with two backticks in reStructuredText.

.. image:: img/spring_arm_pivot_setup.png

> Give the spring arm a length of `3` so that it is placed behind the character:

Reviewer: ```suggestion
Give the spring arm a length of ``3`` so that it is placed behind the character:
```

.. image:: img/spring_arm_length_setup.png

> .. note:: Leave the "Shape" of the spring arm as `<empty>`. This way, it will use the camera's pyramid shape.

Reviewer: ```suggestion
.. note:: Leave the **Shape** of the spring arm as ``<empty>``. This way, it will use the camera's pyramid shape.
```

```suggestion
.. note:: Leave the ``Shape`` of the spring arm as ``<empty>``. This way, it will use the camera's pyramid shape.
```
Pick one of these, either works

    If you want, you can also try other shapes - a sphere is a common choice since it slides smoothly along edges.

> Update the top of `player.gd` to grab the camera and the pivot points by their unique names:

Reviewer: ```suggestion
Update the top of ``player.gd`` to grab the camera and the pivot points by their unique names:
```

> .. code-block:: gdscript

Reviewer: ```suggestion
.. code-block:: gdscript
    :caption: player.gd
```

    @onready var _camera := %Camera3D as Camera3D

>     @onready var _camera_pivot := %CameraPivot as Node3D

Reviewer: Adding this code as-is gives me an error. I also had to remove an existing `_camera` property.

Author: Added a not here to comment out the existing line 👍

> Add an `_unhandled_input` function to check for camera movement and then rotate the pivot point accordingly:

Reviewer: ```suggestion
Add an ``_unhandled_input`` function to check for camera movement and then rotate the pivot point accordingly:
```

> .. code-block:: gdscript
> 

Reviewer: ```suggestion
.. code-block:: gdscript
    :caption: player.gd
```
Add filenames for code in captions (new since https://github.com/godotengine/godot-docs/pull/10151)

    func _unhandled_input(event: InputEvent) -> void:
        if event is InputEventMouseMotion:
            _camera_pivot.rotation.x += event.relative.y * 0.01

>             _camera_pivot.rotation.y += -event.relative.x * 0.01

Reviewer: Similarly, this code didn't work until I deleted the existing camera node in the player scene. I think it was blocking the unhandled input somehow?

You may also want to add a one-liner for vertical clamping code. The result of small tutorials like this doesn't need to be production-ready, but it should avoid obvious errors like allowing 360 degree unclamped motion in a character controller:

https://github.com/user-attachments/assets/8f04d7d0-6c3c-4fc2-885f-2e658fb0800c

Author: Added a note about removing the existing camera and made the screenshot a bit larger so you can see all the nodes in the scene.

Also, good call on the clamping logic 👍

:article_outdated: False

.. _doc_spring_arm:

Third-person camera with spring arm
===================================

Introduction
------------

3D games will often have a third-person camera that follows and
rotates around something such as a player character or a vehicle.

In Godot, this can be done by setting a :ref:`Camera3D <class_Camera3D>` as a child of a node.

> However, if you try this without any extra steps you'll notice that the camera clips through geometry and hides the scene.

Reviewer 2: ```suggestion
However, if you try this without any extra steps, you'll notice that the camera clips through geometry and hides the scene.
```

Reviewer 2: Applied in https://github.com/godotengine/godot-docs/commit/e95419499130b61f5d196d64d7e413795c2e736c.

Every physics process frame, the spring arm will perform a motion cast to check if anything is collided with:

.. image:: img/spring_arm_camera_motion_cast.webp

When the shape hits something, the camera will be placed at or near the collision point:

.. image:: img/spring_arm_camera_collision.webp

Setting up the spring arm and camera
------------------------------------

Let's add a spring arm camera setup to the platformer demo.

.. note:: You can download the Platformer 3D demo on `GitHub <https://github.com/godotengine/godot-demo-projects/tree/master/3d/platformer>`_ or using the `Asset Library <https://godotengine.org/asset-library/asset/2748>`_.

In general, for a third-person camera setup, you will have three nodes as children of the node that you're following:

- `Node3D` (the "pivot point" for the camera)

    - `SpringArm3D`

        - `Camera3D`

Open the ``player/player.tscn`` scene. Set these up as children of our player and give them unique names so we can find them in our script. **Make sure to delete the existing camera node!**

.. image:: img/spring_arm_editor_setup.webp

> Let's move the pivot point up by ``3`` on the Y-axis so that it's not on the ground:

Reviewer: ```suggestion
Let's move the pivot point up by ``2`` on the Y-axis so that it's not on the ground:
```
(And change the screenshot if you can)

Nitpick, but a value of `3` puts the camera pivot solidly above the head of the player:

![Godot_v4 3-stable_win64_fuIwdhkgeo](https://github.com/user-attachments/assets/9458eddb-f170-4277-9b6d-19694f6811ff)

If you stand close to the wall and try to clip the camera into the wall, instead you don't see the player at all:

https://github.com/user-attachments/assets/cb498802-8fb9-49e0-bfc2-77c8256e7b2f

Reviewer: **y = 2**
![Godot_v4 3-stable_win64_RDCBcYZr27](https://github.com/user-attachments/assets/31eadeaa-5a33-4729-95f7-ebc84f1e7d3b)

**y = 3**
![Godot_v4 3-stable_win64_Ap5XPJDF0A](https://github.com/user-attachments/assets/0c4276fe-fde9-4059-a7ad-6e371e9a7161)

Maybe there's some other difference in setup. I imported the demo into 4.3, then made the changes as suggested by your tutorial. The CameraPivot is a direct child of the Player node, and it has a position of either `(0,2,0) or `(0,3,0)`.

This isn't a huge deal either way. I just noticed because I tried to recreate the video at the end of the tutorial, and the pivot was too far off the ground to actually see the player as the camera rotated. I could still see how the spring arm generally works.

Author: Okay, you're totally right! I had some other changes that were affecting it

Author: Okay, updated it to be `2` instead of `3` 👍

.. image:: img/spring_arm_pivot_setup.webp

Give the spring arm a length of ``3`` so that it is placed behind the character:

.. image:: img/spring_arm_length_setup.webp

.. note:: Leave the **Shape** of the spring arm as ``<empty>``. This way, it will use the camera's pyramid shape.

    If you want, you can also try other shapes - a sphere is a common choice since it slides smoothly along edges.

> Update the top of ``player/player.tscn`` to grab the camera and the pivot points by their unique names:

Reviewer: ```suggestion
Update the top of ``player/player.gd`` to grab the camera and the pivot points by their unique names:
```
Editing the script, not the scene.

> .. code-block:: gdscript
>     :caption: player/player.tscn
> 
>     # remove this existing camera line
>     # @onready var _camera := $Target/Camera3D as Camera3D
> 
>     @onready var _camera := %Camera3D as Camera3D
>     @onready var _camera_pivot := %CameraPivot as Node3D

Reviewer: ```suggestion
.. code-block:: gdscript
    :caption: player/player.gd

    # Comment out this existing camera line.
    # @onready var _camera := $Target/Camera3D as Camera3D

    @onready var _camera := %Camera3D as Camera3D
    @onready var _camera_pivot := %CameraPivot as Node3D
```

Add an ``_unhandled_input`` function to check for camera movement and then rotate the pivot point accordingly:

> .. code-block:: gdscript
>     :caption: player/player.tscn
> 
>     @export_group("Camera")
>     @export_range(0.0, 1.0) var mouse_sensitivity := 0.01
>     @export var tilt_upper_limit := PI / 3.0
>     @export var tilt_lower_limit := -PI / 3.0
> 
> 
>     func _unhandled_input(event: InputEvent) -> void:
>         if event is InputEventMouseMotion:
>             _camera_pivot.rotation.x -= event.relative.y * mouse_sensitivity
> 
>             # this will keep the camera from rotating too far up or down
>             _camera_pivot.rotation.x = clampf(
>                 _camera_pivot.rotation.x, tilt_lower_limit, tilt_upper_limit
>             )
> 
>             _camera_pivot.rotation.y += -event.relative.x * mouse_sensitivity
> 

Reviewer: ```suggestion
.. code-block:: gdscript
    :caption: player/player.gd

@export_range(0.0, 1.0) var mouse_sensitivity = 0.01
@export var tilt_limit = deg_to_rad(75)

func _unhandled_input(event: InputEvent) -> void:
	if event is InputEventMouseMotion:
		_camera_pivot.rotation.x -= event.relative.y * mouse_sensitivity
		# Prevent the camera from rotating too far up or down.
		_camera_pivot.rotation.x = clampf(_camera_pivot.rotation.x, -tilt_limit, tilt_limit)
		_camera_pivot.rotation.y += -event.relative.x * mouse_sensitivity
```
Some code style nitpicks:

- We're actually editing the `.gd`, not the scene
- Use dynamic typing. I don't love it either, but it's the [current guideline](https://docs.godotengine.org/en/stable/contributing/documentation/docs_writing_guidelines.html#dynamic-vs-static-typing)
- Export group is a detail that doesn't matter for this example (even if it would be good practice in a larger script)
- Use a single tilt limit
- IMO deg_to_rad on a round multiple of 5 looks cleaner than a multiple of PI. This one's subjective, feel free to ignore
- Comments are sentence case.
- Make the one-liner actually a one-liner
